### ISSUE TITLE
Corretto-8 Caused by: java.security.UnrecoverableKeyException: Password verification failed when not explicitly setting custom truststore

### ISSUE THREAD

**Issue by ryankazokas** [corretto/corretto-8#300]
### Describe the bug
No default truststore in the corretto-8 lambda runtime or is not accessible. When I download the corretto-8 jdk i can see that there is a cacerts in the directory and I have no issues when running outside of lambda connecting to rds via SSL. However when we try to make a call via SSL to rds instance we are seeing the below error.

```
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: Cannot open JKS [Keystore was tampered with, or password was incorrect]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.Util.getInstance(Util.java:408)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:918)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:897)
	at com.mysql.jdbc.ExportControlled.getSSLSocketFactoryDefaultOrConfigured(ExportControlled.java:455)
	at com.mysql.jdbc.ExportControlled.transformSocketToSSLSocket(ExportControlled.java:103)
	at com.mysql.jdbc.MysqlIO.negotiateSSLConnection(MysqlIO.java:4901)
	at com.mysql.jdbc.MysqlIO.proceedHandshakeWithPluggableAuthentication(MysqlIO.java:1659)
	at com.mysql.jdbc.MysqlIO.doHandshake(MysqlIO.java:1226)
	at com.mysql.jdbc.ConnectionImpl.coreConnect(ConnectionImpl.java:2191)
	at com.mysql.jdbc.ConnectionImpl.connectWithRetries(ConnectionImpl.java:2038)
	... 72 common frames omitted
Caused by: java.io.IOException: Keystore was tampered with, or password was incorrect
	at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:792)
	at sun.security.provider.JavaKeyStore$JKS.engineLoad(JavaKeyStore.java:57)
	at sun.security.provider.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:224)
	at sun.security.provider.JavaKeyStore$DualFormatJKS.engineLoad(JavaKeyStore.java:71)
	at java.security.KeyStore.load(KeyStore.java:1445)
	at com.mysql.jdbc.ExportControlled.getSSLSocketFactoryDefaultOrConfigured(ExportControlled.java:428)
	... 78 common frames omitted
Caused by: java.security.UnrecoverableKeyException: Password verification failed
	at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:790)
```
	
We've just copied the keystore and are setting in code now where to look for it and that seems to work. Is there a way to leverage the cacerts that comes prebuilt with the SDK in a lambda environment, so that we can make the SSL calls. This works for us for now, but is not a scalable solution.

```
		final File jks = new File("cacerts2");
		System.setProperty("javax.net.ssl.trustStore", jks.getAbsolutePath());
```

### To Reproduce
Make a call to an RDS instance via jdbc MySQL driver with a jdbc connection string that has the following flags set to true: useSsl=true and requireSsl=tru

### Expected behavior
I'd expect this to work out of the box if possible. Or at least be able to specify in code, the truststore that comes with the corretto-8 jdk.

### Platform information
    OS: corretto-8 lambda built with and deployed with SAM template.
    Runtime: java8.al2

**Comment by alvdavi** [comment]
Hi,

Lambda does include a copy of the default cacerts. However, seeing that exception reminds me of an old issue in how OpenJDK8 manages the fallback to using the default cacerts.

Whenever the `javax.net.ssl.trustStore` property is set, OpenJDK will attempt to open the keystore specified. Additionally, a `javax.net.ssl.trustStorePassword` can be provided. If for any reason the JDK fails to open the specified keystore, it will attempt to fall back to the default cacerts. However, if a password was provided, it will attempt to use that password to decrypt the default cacerts (that has no password), causing the `Password verification failed` exception you have.

This is a copy of the exception I recorded when analyzing this, as you can see, it is practically identical to the one you posted:
```
Caused by: java.io.IOException: Keystore was tampered with, or password was incorrect
    at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:785)
    at sun.security.provider.JavaKeyStore$JKS.engineLoad(JavaKeyStore.java:56)
    at sun.security.provider.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:224)
    at sun.security.provider.JavaKeyStore$DualFormatJKS.engineLoad(JavaKeyStore.java:70)
    at java.security.KeyStore.load(KeyStore.java:1445)
    at sun.security.ssl.TrustStoreManager$TrustAnchorManager.loadKeyStore(TrustStoreManager.java:367)
    at sun.security.ssl.TrustStoreManager$TrustAnchorManager.getTrustedCerts(TrustStoreManager.java:315)
    at sun.security.ssl.TrustStoreManager.getTrustedCerts(TrustStoreManager.java:59)
    at sun.security.ssl.TrustManagerFactoryImpl.engineInit(TrustManagerFactoryImpl.java:51)
    ... 44 more
Caused by: java.security.UnrecoverableKeyException: Password verification failed
    at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:783)
```

You might have some more information on the email I wrote back then:
https://mail.openjdk.java.net/pipermail/jdk8u-dev/2019-August/009999.html

My suggestion, if you want to make use of the default cacerts file, make sure your code is not setting the `javax.net.ssl.trustStore` and/or `javax.net.ssl.trustStorePassword`. If you still have an issue, I suggest looking at the value of `javax.net.ssl.trustStore`, `javax.net.ssl.trustStorePassword` and `user.dir` when the exception is thrown.

**Comment by ryankazokas** [comment]
Thanks for the response @alvdavi . It does look like it's picking the trust store on the lambda, however the password is null by default. I'm guessing that might be the root cause of my issue.  

The following env vars for user.dir, javax.net.ssl.trustStore, javax.net.ssl.trustStorePassword
```
2021-04-27 14:43:30.339  INFO   --- [           main] com.x.JerseyRequestHandler      : /var/task
2021-04-27 14:43:30.340  INFO   --- [           main] com.x.JerseyRequestHandler      : /etc/pki/java/cacerts
2021-04-27 14:43:30.340  INFO   --- [           main] com.x.JerseyRequestHandler      : null
```

Is there a way to get / set the password here? or should it be null? One other thing I tried was just setting the password as empty string but the same thing happens here. Sorry if I'm missing something obvious.

**Comment by alvdavi** [comment]
Hi, the default cacerts does not have a password, so it is ok for it to be null. My doubt seeing those values is when where they captured? Were those values printed when the exception was thrown, or before? It is possible for those properties to have one value at the beginning but be changed later on.

Finally, why the need of passing the `javax.net.ssl.trustStore`? What will be the result if you avoid setting these properties at all?

**Comment by ryankazokas** [comment]
@alvdavi it was the same when it got to the point of logging, but I found the real issue. It looks like due to some dependencies in our parent pom(which was a dependant of spring-boot 1.5.8), we were using an outdated version of the MySQL driver. 5.x.x. After discovering on a stackoverflow post, that this particular warning log was fixed in MySQL java driver 8.0.16
https://dev.mysql.com/doc/relnotes/connector-j/8.0/en/news-8-0-16.html

Thanks for the guidance on troubleshooting. I'm going to close this out.
